Clean up and encourage immutability within the codebase

jbduncan opened the pull request:
## Overview

I've made a number of cleanup- and immutability-related changes to the JUnit 5 codebase in my spare time, mainly for fun, but also in the hope that it makes things a little easier to maintain in the future. However, it's not clear to me if all the changes I've made are actually desirable for backwards-compatibility or other reasons.

Are all the changes I've made here OK? If not, I wonder if someone on the JUnit 5 team would kindly let me know either where I can improve or where I've overstepped my boundaries. :)

Here's a summary of the main changes I've made, as I've written in my commit message:
```
    Clean up and encourage immutability within the codebase

    - `Collections::unmodifiable*` has been used more for the return values
      of public, non-internal, collection-returning methods, with the
      intention of making them a bit easier to use and understand for
      external users (since their mutability (or lack thereof) is now more
      clearly defined and enforced).
    - Code blocks between `// formatter:{off|on}` comments have been
      manually formatted for consistenty.
    - The methods of `Collectors` are statically imported more for
      consistency.
```

---

I hereby agree to the terms of the JUnit Contributor License Agreement.

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests)
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [ ] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

jbduncan commented:
P.S. I've noted that the automatic checks have failed, so I'm looking into it.

jbduncan commented:
Looking back over this, I've realised that this is quite a large PR, so if it's too much to go through at once, I'd be more than happy to close it and submit a few smaller, broken-down ones.

marcphilipp commented:
It's okay IMHO. We just didn't have time to review it, yet. Sorry for the delay!

sormuras commented on junit-platform-commons/src/main/java/org/junit/platform/commons/util/JUnitCollectors.java:
```diff
@@ -0,0 +1,66 @@
+ * @since 1.0
+ */
+@API(Internal)
+public final class JUnitCollectors {
```
Rename to `CollectorUtils`?

jbduncan commented:
No worries @marcphilipp. :)

jbduncan commented:
@sormuras:

> Move toUnmodifiableList() to existing helper class CollectionUtils and remove JUnitCollectors class. If we have many collector factories some day in the future, we could introduce a dedicated CollectorUtils helper then.

Sounds good to me! Working on it now.

jbduncan commented:
Done, I've made a little commit at https://github.com/junit-team/junit5/pull/705/commits/5822313edf9478fe89cf9eb7a59a90f20a361018 which I'll squash into the first one when the review's finished.

sormuras commented on junit-platform-launcher/src/main/java/org/junit/platform/launcher/listeners/MutableTestExecutionSummary.java:
```diff
@@ -192,13 +194,13 @@ public void printFailuresTo(PrintWriter writer) {
 	@Override
 	public List<Failure> getFailures() {
-		return new ArrayList<>(failures);
+		return Collections.unmodifiableList(new ArrayList<>(failures));
```
`return Collections.unmodifiableList(failures);` should suffice, iirc.

jbduncan replied:
I'm not so sure about that, because it would change the behaviour of `getFailures()` to not only be unmodifiable but also a _view_ rather than a copy, and I don't know if that behaviour is relied upon. (But then again, I don't know if `getFailures()` being mutable is relied upon as well...)

jbduncan replied:
Since JUnit 5 does not have a 1.0 release yet, would it be alright to modify both the mutability and view/copy-ness of `getFailures()` as long as it's recorded in the release notes?

sormuras replied:
Imho, yes. "JUnit 5" is the only "client" affected by the change, anyway. `MutableTestExecutionSummary` is package private and therefore `@API(Internal)`.

jbduncan replied:
Okay, sounds good to me. Implementing change now.

sormuras commented on junit-platform-commons/src/main/java/org/junit/platform/commons/util/AnnotationUtils.java:
```diff
@@ -178,7 +178,8 @@ public static boolean isAnnotated(AnnotatedElement element, Class<? extends Anno
 		findRepeatableAnnotations(element, annotationType, containerType, inherited, found, new HashSet<>(16));
-		return new ArrayList<>(found);
+		// unmodifiable since returned from public, non-internal method(s)
+		return Collections.unmodifiableList(new ArrayList<>(found));
```
`return Collections.unmodifiableList(found);` should suffice, iirc. ;)

jbduncan commented:
Well, third attempt to (hopefully) rectify CI failures, as well as add a rudimentary test for `CollectionUtils::toUnmodifiableList`, and to fix some other things. (https://github.com/junit-team/junit5/pull/705/commits/69ce8adde0e41248feb0518191b6ce0b6c73c5bb)

jbduncan commented:
An alternative to the rudimentary test that I wrote for `CollectionUtils::toUnmodifiableList`, which would be more resilient, is to import guava-testlib and use its `CollectorTester`.

Thoughts?

jbduncan commented:
Oops, just a _little_ problem :wink::
guava-testlib's `ListTestSuiteBuilder` (which I'd like to use alongside `CollectorTester`, since it isn't as powerful as I thought), returns JUnit 3 `TestSuite`s, and it's not really clear to me how to get JUnit 5 to run them if possible.

Would the JUnit 5 team be able to point me in the right direction? :smile:

marcphilipp commented:
Please don't go overboard with testing the collector. It's not our own implementation, so IMHO it would suffice to test that the return list is indeed immutable.

jbduncan commented:
Okey dokey @marcphilipp, that makes reasonable sense to me. :)

jbduncan commented:
@sormuras @marcphilipp I've made a change as requested in https://github.com/junit-team/junit5/pull/705/files/11f3dfbe91a418be7635d26a3a6db04dd5c4e340#r104801114, and I've rebased and squashed all the commits into one now, ready for a 2nd review.

gaganis commented on junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/DiscoverySelectors.java:
```diff
@@ -46,7 +46,8 @@
 public final class DiscoverySelectors {
 
-	private static Pattern fullyQualifiedMethodNamePattern = Pattern.compile("([^#]+)#([^(]+)(?:\\((.*)\\))?");
+	private static final Pattern FULLY_QUALIFIED_METHOD_NAME_PATTERN = Pattern.compile(
```
This is good catch, converting this to an idiomatic constant. Is this just a refactoring or is this part of improving immutability? 

Thanks!
Giorgos

jbduncan replied:
@gaganis It was just a refactoring into an idiomatic constant, although the `final` does help with immutability. :)

jbduncan commented:
Hrm, this PR now has conflicts with `junit5-team/junit5:master`, and I've been trying to find a way of rebasing `jbduncan/junit5:clean-up-and-immutability` over it to resolve the conflicts, but with little success so far. (I want to avoid merging branches if I can, to not pollute the history.)

Any ideas?

gaganis commented:
@jbduncan You can still rebase and resolve your conflicts. This can be done during the rebase operation.

Unfortunately resolving conflicts is something that cannot be done automatically, you will have to edit the conflicts manually. -This is also true if you tried to use a merge commit instead of rebasing-. 

Github help provides a guide on how to do a rebase with conflicts:
https://help.github.com/articles/resolving-merge-conflicts-after-a-git-rebase/

They also provide tips on how to do the manual conflict resolution in this:
https://help.github.com/articles/resolving-a-merge-conflict-using-the-command-line/

If you get stuck and need any tips feel free to ping me :)

jbduncan commented:
Thanks for the advice @gaganis, but I realise that I was unclear in what I was asking for.

I actually already knew how to resolve merge conflicts, but I was struggling to find out how to even rebase `jbduncan/junit5:clean-up-and-immutability` over `junit5-team/junit5:master` in the first place, so that I could manually resolve them.

Thankfully I found this blog on [keeping a Github fork updated](https://robots.thoughtbot.com/keeping-a-github-fork-updated) which solved my problem, so everything's fine now. :)

jbduncan commented:
Ready for a 2nd review now @sormuras.

sormuras commented:
Almost. (-: See AppVeyor.

Am 11.03.2017 4:28 nachm. schrieb "Jonathan Bluett-Duncan" <
notifications@github.com>:

> Ready for a 2nd review now @sormuras <https://github.com/sormuras>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/junit-team/junit5/pull/705#issuecomment-285874896>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACNl3tiFEWsrfGrZ81SRA2Z7VuOYBnKfks5rkr2jgaJpZM4MTgiG>
> .
>

jbduncan commented:
Good catch @sormuras! Fixed. :)

sormuras commented:
Thanks @jbduncan
